crypto/crypto/src/noise.rs

```diff
@@ -0,0 +1,587 @@
+      rng.fill_bytes(&mut ephemeral_private);
+      x25519::StaticSecret::from(ephemeral_private)
+    } else {
+      x25519::StaticSecret::new(rng)
+    };
+    let e_pub = x25519::PublicKey::from(&e);
+    mix_hash(&mut h, e_pub.as_bytes());
+    let mut msg = e_pub.as_bytes().to_vec();
+
+    // -> ee
+    let dh_output = e.diffie_hellman(&remote_ephemeral);
+    mix_key(&mut ck, dh_output.as_bytes())?;
+
+    // -> se
+    let dh_output = e.diffie_hellman(&remote_static);
+    let k = mix_key(&mut ck, dh_output.as_bytes())?;
+
+    // -> payload
+    #[cfg(test)]
+    let aead = Aes256Gcm::new(GenericArray::clone_from_slice(&k));
+    #[cfg(not(test))]
+    let aead = Aes128Gcm::new(GenericArray::clone_from_slice(&k));
+
+    let msg_and_ad = Payload {
+      msg: payload.unwrap_or_else(|| &[]),
+      aad: &h,
+    };
+    let nonce = GenericArray::from_slice(&[0u8; 12]);
+    let encrypted_payload = aead
+      .encrypt(nonce, msg_and_ad)
+      .map_err(|_| NoiseError::Encrypt)?;
+    mix_hash(&mut h, &encrypted_payload);
+    msg.extend_from_slice(&encrypted_payload);
+
+    // split
+    let (mut k1, mut k2) = hkdf(&ck, None)?;
+    if !cfg!(test) {
+      k1.resize(16, 0);
+      k2.resize(16, 0);
+    }
+    let session = NoiseSession::new(k2, k1);
+
+    //
+    Ok((msg, remote_static, received_payload, session))
+  }
+}
+
+//
+// Post-Handshake (NoiseSession)
+// ------------------------
+
+/// A NoiseSession is produced after a successful Noise handshake, and can be use to encrypt and decrypt messages to the other peer.
+pub struct NoiseSession {
+  /// a session can be marked as invalid if it has seen a decryption failure
+  valid: bool,
+  /// key used to encrypt messages to the other peer
+  write_key: Vec<u8>,
+  /// associated nonce (in practice the maximum u64 value cannot be reached)
+  write_nonce: u64,
+  /// key used to decrypt messages received from the other peer
+  read_key: Vec<u8>,
+  /// associated nonce (in practice the maximum u64 value cannot be reached)E
+  read_nonce: u64,
+}
+
+impl NoiseSession {
+  fn new(write_key: Vec<u8>, read_key: Vec<u8>) -> Self {
+    Self {
+      valid: true,
+      write_key,
+      write_nonce: 0,
+      read_key,
+      read_nonce: 0,
+    }
+  }
+
+  /// encrypts a message for the other peers (post-handshake)
+  pub fn write_message(&mut self, plaintext: &[u8]) -> Result<Vec<u8>, NoiseError> {
+    // checks
+    if !self.valid {
+      return Err(NoiseError::SessionClosed);
+    }
+    if plaintext.len() > MAX_SIZE_NOISE_MSG - AES_GCM_TAGLEN {
+      return Err(NoiseError::PayloadTooLarge);
+    }
+
+    // encrypt
+    #[cfg(test)]
+    let aead = Aes256Gcm::new(GenericArray::clone_from_slice(&self.write_key));
+    #[cfg(not(test))]
+    let aead = Aes128Gcm::new(GenericArray::clone_from_slice(&self.write_key));
+
+    let msg_and_ad = Payload {
+      msg: plaintext,
+      aad: &[],
+    };
+    let mut nonce = [0u8; 4].to_vec();
```

Comment:
Why do we have a fixed nonce starting from zero?

Reply:
that's the spec

Reply:
(to add to this, Noise uses this key to encrypt random data + only use the key once)